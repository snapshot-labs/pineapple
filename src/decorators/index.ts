export {
  withServiceMetrics,
  type MetricsContext,
  type Service,
  type ServiceMethod
} from './withServiceMetrics';
