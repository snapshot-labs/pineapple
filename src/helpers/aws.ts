import { Readable } from 'stream';
import * as AWS from '@aws-sdk/client-s3';

let client;
const region = process.env.AWS_REGION;
const endpoint = process.env.AWS_ENDPOINT || undefined;
if (region) client = new AWS.S3({ region, endpoint });
const dir = 'pineapple';

async function streamToString(stream: Readable): Promise<string> {
  return await new Promise((resolve, reject) => {
    const chunks: Uint8Array[] = [];
    stream.on('data', chunk => chunks.push(chunk));
    stream.on('error', reject);
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf-8')));
  });
}

export async function set(key, value) {
  if (!client) return;
  try {
    return await client.putObject({
      Bucket: process.env.AWS_BUCKET_NAME,
      Key: `public/${dir}/${key}`,
      Body: JSON.stringify(value),
      ContentType: 'application/json; charset=utf-8'
    });
  } catch (e) {
    console.log('Store cache failed', e);
    throw e;
  }
}

export async function get(key) {
  if (!client) return false;
  try {
    const { Body } = await client.getObject({
      Bucket: process.env.AWS_BUCKET_NAME,
      Key: `public/${dir}/${key}`
    });
    const str = await streamToString(Body);
    return JSON.parse(str);
  } catch (e) {
    return false;
  }
}

export async function remove(key) {
  if (!client) return false;
  try {
    return await client.deleteObject({
      Bucket: process.env.AWS_BUCKET_NAME,
      Key: `public/${dir}/${key}`
    });
  } catch (e: any) {
    return false;
  }
}
