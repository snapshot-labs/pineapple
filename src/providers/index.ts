import { capture } from '@snapshot-labs/snapshot-sentry';
import ipfsProviders, {
  IMAGE_PROVIDERS as IPFS_IMAGE_PROVIDERS,
  JSON_PROVIDERS as IPFS_JSON_PROVIDERS
} from './ipfs';
import { Protocol, ProviderMap, UploadOptions } from './types';
import {
  countOpenProvidersRequest,
  providersUploadSize,
  timeProvidersUpload
} from '../helpers/metrics';

const PROVIDERS = {
  ipfs: {
    image: IPFS_IMAGE_PROVIDERS,
    json: IPFS_JSON_PROVIDERS,
    list: ipfsProviders as ProviderMap
  }
};

export const DEFAULT_PROTOCOL: Protocol = 'ipfs';

export default function uploadToProviders(
  options: UploadOptions
): Promise<{ cid: string; provider: string }> {
  const { protocol, type, params, customProviderIds } = options;
  const providerIds = customProviderIds || PROVIDERS[protocol][type];
  const configuredProviders = providerIds
    .map(p => PROVIDERS[protocol].list[p])
    .filter(p => p?.isConfigured());

  return Promise.any(
    configuredProviders.map(async ({ id, set }) => {
      const end = timeProvidersUpload.startTimer({ name: id, type });
      let status = 0;

      try {
        countOpenProvidersRequest.inc({ name: id, type });

        const result = await set(params);
        const size = (params instanceof Buffer ? params : Buffer.from(JSON.stringify(params)))
          .length;
        providersUploadSize.inc({ name: id, type }, size);
        status = 1;

        return result;
      } catch (e: any) {
        if (e instanceof Error) {
          if (e.message !== 'Request timed out') {
            capture(e, { name: id });
          }
        } else {
          capture(new Error(`Error from ${id} provider`), {
            contexts: { provider_response: e }
          });
        }
        return Promise.reject(e);
      } finally {
        end({ status });
        countOpenProvidersRequest.dec({ name: id, type });
      }
    })
  );
}
