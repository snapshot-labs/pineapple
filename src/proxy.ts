import express from 'express';
import fetch from 'cross-fetch';
import Promise from 'bluebird';
import gateways from './gateways.json';
import { set, get } from './aws';
import { MAX, sha256 } from './utils';

const router = express.Router();

router.get('/ipfs/*', async (req, res) => {
  const key = sha256(req.originalUrl);
  try {
    const cache = await get(`cache/${key}`);
    if (cache) return res.json(cache);
    const json = await Promise.any(
      gateways.map((gateway) => {
        const url = `https://${gateway}${req.originalUrl}`;
        return fetch(url).then((res) => res.json());
      })
    );
    res.json(json);
    const size = Buffer.from(JSON.stringify(json)).length;
    if (size <= MAX) await set(`cache/${key}`, json);
  } catch (e) {
    return res.status(500).json();
  }
});

export default router;
