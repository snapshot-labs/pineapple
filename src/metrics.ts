import init, { client } from '@snapshot-labs/snapshot-metrics';
import { capture } from '@snapshot-labs/snapshot-sentry';
import { Express } from 'express';
import gateways from './gateways.json';
import { IMAGE_PROVIDERS, JSON_PROVIDERS, providersMap } from './providers/utils';

export default function initMetrics(app: Express) {
  init(app, {
    normalizedPath: [['^/ipfs/.*', '/ipfs/#url']],
    whitelistedPath: [/^\/$/, /^\/upload$/, /^\/ipfs\/.*$/],
    errorHandler: capture
  });

  app.use(providersInstrumentation);
}

const gatewaysCount = new client.Gauge({
  name: 'ipfs_gateways_count',
  help: 'Number of IPFS gateways.'
});
gatewaysCount.set(gateways.length);

const providersJsonCount = new client.Gauge({
  name: 'providers_json_count',
  help: 'Number of providers used for JSON pinning.'
});
providersJsonCount.set(JSON_PROVIDERS.filter(p => providersMap[p].isConfigured()).length);

const providersImageCount = new client.Gauge({
  name: 'providers_image_count',
  help: 'Number of providers used for image pinning.'
});
providersImageCount.set(IMAGE_PROVIDERS.filter(p => providersMap[p].isConfigured()).length);

export const timeProvidersUpload = new client.Histogram({
  name: 'providers_upload_duration_seconds',
  help: "Duration in seconds of provider's upload requests.",
  labelNames: ['name', 'type', 'status'],
  buckets: [0.5, 1, 2, 5, 10, 15]
});

export const providersUploadSize = new client.Counter({
  name: 'providers_upload_size',
  help: "Total size of each provider's upload file.",
  labelNames: ['name', 'type']
});

const providersReturnCount = new client.Counter({
  name: 'providers_return_count',
  help: 'Number of times each provider have been used.',
  labelNames: ['name', 'type']
});

export const timeIpfsGatewaysResponse = new client.Histogram({
  name: 'ipfs_gateways_response_duration_seconds',
  help: "Duration in seconds of each IPFS gateway's reponse.",
  labelNames: ['name', 'status'],
  buckets: [0.5, 1, 2, 5, 10, 15]
});

export const ipfsGatewaysReturnCount = new client.Counter({
  name: 'ipfs_gateways_return_count',
  help: 'Number of times each gateway have been used.',
  labelNames: ['name']
});

export const ipfsGatewaysCacheHitCount = new client.Counter({
  name: 'ipfs_gateways_cache_hit_count',
  help: 'Number of hit/miss of the IPFS gateways cache layer',
  labelNames: ['status']
});

export const ipfsGatewaysCacheSize = new client.Counter({
  name: 'ipfs_gateways_cache_size',
  help: 'Total size going through the IPFS gateways cache layer',
  labelNames: ['status']
});

export const countOpenProvidersRequest = new client.Gauge({
  name: 'providers_open_connections_count',
  help: 'Number of open connections to providers.',
  labelNames: ['name', 'type']
});

export const countOpenGatewaysRequest = new client.Gauge({
  name: 'ipfs_gateways_open_connections_count',
  help: 'Number of open connections to gateways.',
  labelNames: ['name']
});

const providersInstrumentation = (req, res, next) => {
  let type;
  if (req.method === 'POST' && req.originalUrl === '/') {
    type = 'json';
  } else if (req.method === 'POST' && req.originalUrl === '/upload') {
    type = 'image';
  }

  if (!type) {
    return next();
  }

  const oldJson = res.json;
  res.json = body => {
    if (res.statusCode === 200 && body) {
      providersReturnCount.inc({ name: body.result?.provider || body.provider, type });
    }

    res.locals.body = body;
    return oldJson.call(res, body);
  };
  next();
};
