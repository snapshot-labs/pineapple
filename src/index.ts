import 'dotenv/config';
import { fallbackLogger, initLogger } from '@snapshot-labs/snapshot-sentry';
import compression from 'compression';
import cors from 'cors';
import express from 'express';
import initMetrics from './helpers/metrics';
import proxy from './routes/proxy';
import uploadImage from './routes/uploadImage';
import uploadJson from './routes/uploadJson';
import { version } from '../package.json';

const app = express();
const PORT = process.env.PORT || 3000;

const commit = process.env.COMMIT_HASH || '';
const v = commit ? `${version}#${commit.substr(0, 7)}` : version;

initLogger(app);
initMetrics(app);

app.disable('x-powered-by');
app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ limit: '4mb', extended: false }));
app.use(cors({ maxAge: 86400 }));
app.use(compression());
app.use('/', uploadJson);
app.use('/', uploadImage);
app.use('/', proxy);
app.get('/', (req, res) => res.json({ version: v, port: PORT }));

fallbackLogger(app);

app.listen(PORT, () => console.log(`Listening at http://localhost:${PORT}`));
