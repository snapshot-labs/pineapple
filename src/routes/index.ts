import { Express } from 'express';
import proxy from './proxy';
import rpc from './rpc';
import upload from './upload';
import { version } from '../../package.json';

export default function setupRoutes(app: Express): void {
  const PORT = process.env.PORT || 3000;
  const commit = process.env.COMMIT_HASH || '';
  const v = commit ? `${version}#${commit.substr(0, 7)}` : version;

  app.use('/', rpc);
  app.use('/', upload);
  app.use('/', proxy);
  app.get('/', (req, res) => res.json({ version: v, port: PORT }));
}
