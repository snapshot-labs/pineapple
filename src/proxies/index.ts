import ipfsProxies from './ipfs';
import { Protocol, ProxiesMap, Response } from './types';
import { countOpenProxyRequest, proxyReturnCount, timeProxyResponse } from '../helpers/metrics';

export const UNSUPPORTED_FILE_TYPE_ERROR = 'unsupported file type';

const PROXIES: Record<Protocol, ProxiesMap> = {
  ipfs: ipfsProxies
};

export function resolveFromProxies(protocol: Protocol, hash: string): Promise<Response> {
  const proxies = Object.values(PROXIES[protocol]);

  const allPromises = proxies.flatMap(proxy => {
    const resolveResult = proxy.resolve(hash);

    // Handle both single promise and array of promises
    const promises = Array.isArray(resolveResult) ? resolveResult : [resolveResult];

    return promises.map(async promise => {
      const end = timeProxyResponse.startTimer({ name: proxy.id });
      let status = 0;

      try {
        countOpenProxyRequest.inc({ name: proxy.id });

        const result = await promise;
        status = 1;

        proxyReturnCount.inc({ name: proxy.id });

        return result;
      } finally {
        end({ status });
        countOpenProxyRequest.dec({ name: proxy.id });
      }
    });
  });

  return Promise.any(allPromises);
}
