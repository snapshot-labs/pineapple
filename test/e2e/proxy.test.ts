import request from 'supertest';
import { set, get, remove } from '../../src/aws';

const HOST = `http://localhost:${process.env.PORT || 3003}`;

describe('GET /ipfs/:cid', () => {
  describe('when the IPFS cid exists', () => {
    const cid = 'bafkreib5epjzumf3omr7rth5mtcsz4ugcoh3ut4d46hx5xhwm4b3pqr2vi';
    const path = `/ipfs/${cid}`;
    const content = { status: 'OK' };

    afterEach(async () => {
      await remove(cid);
    });

    describe('when the file is cached', () => {
      if (process.env.AWS_REGION) {
        const cachedContent = { status: 'CACHED' };

        it('returns the cache file', async () => {
          await set(cid, cachedContent);
          const response = await request(HOST).get(path);

          expect(response.body).toEqual(cachedContent);
          expect(response.statusCode).toBe(200);
          expect(response.headers['content-type']).toBe('application/json; charset=utf-8');
          expect(await get(cid)).toEqual(cachedContent);
        });
      } else {
        it.todo('needs to set AWS credentials to test the cache');
      }
    });

    describe('when the file is not cached', () => {
      if (process.env.AWS_REGION) {
        it('returns the file and caches it', async () => {
          const response = await request(HOST).get(path);

          expect(response.body).toEqual(content);
          expect(response.statusCode).toBe(200);
          expect(response.headers['content-type']).toBe('application/json; charset=utf-8');
          expect(await get(cid)).toEqual(response.body);
        });
      } else {
        it.todo('needs to set AWS credentials to test the cache');
      }
    });

    it('returns a 415 error when not a JSON file', async () => {
      const response = await request(HOST).get(
        '/ipfs/bafybeie2x4ptheqskiauhfz4w4pbq7o6742oupitganczhjanvffp2spti'
      );

      expect(response.statusCode).toBe(415);
    }, 30e3);
  });

  describe('when the IPFS cid does not exist', () => {
    it('returns a 400 error', async () => {
      const response = await request(HOST).get('/ipfs/test');

      expect(response.statusCode).toBe(400);
    }, 30e3);
  });
});
