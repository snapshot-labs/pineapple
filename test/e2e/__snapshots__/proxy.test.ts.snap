// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /ipfs/:cid when the graph IPFS cid exists returns the file 1`] = `
{
  "tree": [
    {
      "address": "0x91FD2c8d24767db4Ece7069AA27832ffaf8590f3",
      "votingPower": "1",
    },
  ],
}
`;
